src/rest-devices.c

```diff
@@ -0,0 +1,575 @@
+            return -1;
+        }
+        rest_list_add(extension, entry);
+    }
+
+    rest_list_extend(list, extension);
+    return 0;
+}
+
+static int rest_devices_update_list(rest_list_t *list, json_t *jdevice)
+{
+    const char *string;
+    json_t *jstring;
+    rest_list_entry_t *device_entry;
+    database_entry_t *device_data;
+    uint8_t *oldpsk, *oldid;
+    uint8_t binary_buffer[512];
+    size_t length;
+
+    jstring = json_object_get(jdevice, "uuid");
+    string = json_string_value(jstring);
+
+    for (device_entry = list->head; device_entry != NULL; device_entry = device_entry->next)
+    {
+        device_data = (database_entry_t *)device_entry->data;
+
+        if (strcmp(device_data->uuid, string) == 0)
+        {
+            jstring = json_object_get(jdevice, "psk");
+            string = json_string_value(jstring);
+
+            base64_decode(string, NULL, &length);
+            if (base64_decode(string, binary_buffer, &length))
+            {
+                return -1;
+            }
+
+            oldpsk = device_data->psk;
+            device_data->psk = malloc(length);
+            if (device_data->psk == NULL)
+            {
+                device_data->psk = oldpsk;
+                return -1;
+            }
+            memcpy(device_data->psk, binary_buffer, length);
+
+            jstring = json_object_get(jdevice, "psk_id");
+            string = json_string_value(jstring);
+
+            base64_decode(string, NULL, &length);
+            if (base64_decode(string, binary_buffer, &length))
+            {
+                return -1;
+            }
+
+            oldid = device_data->psk_id;
+            device_data->psk_id = malloc(length);
+            if (device_data->psk_id == NULL)
+            {
+                free(device_data->psk);
+                device_data->psk = oldpsk;
+                device_data->psk_id = oldid;
+                return -1;
+            }
+            memcpy(device_data->psk_id, binary_buffer, length);
+            free(oldpsk);
+            free(oldid);
+
+            return 0;
+        }
+    }
+
+    return -1;
+}
+
+static int rest_devices_remove_list(rest_list_t *list, const char *id)
+{
+    rest_list_entry_t *device_entry;
+    database_entry_t *device_data;
+
+    for (device_entry = list->head; device_entry != NULL; device_entry = device_entry->next)
+    {
+        device_data = (database_entry_t *)device_entry->data;
+
+        if (strcmp(id, device_data->uuid) == 0)
+        {
+            rest_list_remove(list, (void *)device_data);
+            return 0;
+        }
+    }
+
+    return -1;
+}
+
+int rest_devices_get_cb(const ulfius_req_t *req, ulfius_resp_t *resp, void *context)
+{
+    rest_context_t *rest = (rest_context_t *)context;
+    rest_list_t *device_list = rest->devicesList;
+    char string[512];
+    size_t length;
+    database_entry_t *device_data;
+    rest_list_entry_t *device_entry;
+    json_t *jdevices = NULL;
```

Comment:
Why did you initialized to `NULL` on declaration?

Reply:
Because callback functions have a single exit path with json_decref(jdevice/jdatabase) etc. In some cases it is possible that a callback will go to exit path without initializing this pointer which could cause a fault.